### ISSUE TITLE
Difference in outputs with dtype `bf16` with `torch.compile`

### ISSUE THREAD

**Issue by shivam15s** [pytorch/pytorch#150638]
### 🐛 Describe the bug

Difference in outputs with torch compile vs eager pytorch
Also. note, setting bias to zeros makes the torch.allclose pass.
Reproducer:

```python
import torch
import torch.nn.functional as F

# seed
torch.manual_seed(0)

def get_per_token_logps(input, weight, bias, selected_token_ids):
    logits = torch.matmul(input, weight.t())
    logits = logits + bias
    logps = F.log_softmax(logits.float(), dim=-1)
    per_token_logps = logps.gather(dim=-1, index=selected_token_ids.unsqueeze(-1)).squeeze(-1)
    return per_token_logps


if __name__ == "__main__":
    B, T, H, V = 8, 128, 1024, 4096
    input = torch.randn(B, T, H, device="cuda", dtype=torch.bfloat16)
    weight = torch.randn(V, H, device="cuda", dtype=torch.bfloat16)
    bias = torch.randn(V, device="cuda", dtype=torch.bfloat16)
    # bias = torch.zeros_like(bias)
    selected_token_ids = torch.randint(0, V, (B, T), device="cuda")
    get_per_token_logps_compiled = torch.compile(get_per_token_logps)
    out = get_per_token_logps(input, weight, bias, selected_token_ids)
    out_compiled = get_per_token_logps_compiled(input, weight, bias, selected_token_ids)
    print(out)
    print(out_compiled)
    print(torch.allclose(out, out_compiled, atol=1e-3, rtol=1e-3))

```

### Error logs

`torch.allclose` returns False

### Versions

```
PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov  6 2024, 20:22:13) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.8.0-1017-azure-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA H100 NVL
GPU 1: NVIDIA H100 NVL

Nvidia driver version: 550.127.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        48 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               80
On-line CPU(s) list:                  0-79
Vendor ID:                            AuthenticAMD
Model name:                           AMD EPYC 9V84 96-Core Processor
CPU family:                           25
Model:                                17
Thread(s) per core:                   1
Core(s) per socket:                   80
Socket(s):                            1
Stepping:                             1
BogoMIPS:                             4800.07
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl tsc_reliable nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext perfctr_core vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves user_shstk avx512_bf16 clzero xsaveerptr rdpru arat avx512vbmi umip avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid fsrm
Hypervisor vendor:                    Microsoft
Virtualization type:                  full
L1d cache:                            2.5 MiB (80 instances)
L1i cache:                            2.5 MiB (80 instances)
L2 cache:                             80 MiB (80 instances)
L3 cache:                             320 MiB (10 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-39
NUMA node1 CPU(s):                    40-79
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Vulnerable: Safe RET, no microcode
Vulnerability Spec store bypass:      Vulnerable
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] flake8==7.1.2
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.2.3
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] torch==2.6.0
[pip3] torch-tb-profiler==0.4.3
[pip3] triton==3.2.0
[conda] No relevant packages
```

**Comment by shivam15s** [comment]
Found `torch.compile` performs `logits = logits + bias` in `float32` while eager does this in `bf16` 🤔 
Is this expected?

**Comment by xmfan** [comment]
Yes, there can be differences in dtype used for intermediate computations. In general, we aim to be close to the fp64 result of the model

**Comment by shivam15s** [comment]
Thanks for your reply. Quick follow up questions:

Doesn’t it cause the code to be slower/consume more memory? Is there a way to ensure dtypes are preserved?

**Comment by xmfan** [comment]
Looking at the add operation in isolation, I think it is a bit slower, but this upcasting only happens when the op is fused with other fp32 tensors. So overall, the fusion saves a few upcasts and downcasts that would have happened under the hood in eager via type promotion.

To get same precision as eager, you can use this flag, but it is mainly used for debugging, not for performance: https://github.com/pytorch/pytorch/blob/d6887f444fa61b46c9c31114028484e658f3dc99/torch/_inductor/config.py#L596-L606
